packages/config/src/class-config.php

```diff
@@ -71,6 +71,38 @@ public function ensure( $feature, array $options = array() ) {
 		$this->set_feature_options( $feature, $options );
 	}
 
+	/**
+	 * Require multiple features to be initialized. It's up to the package consumer to actually add
+	 * the package to their composer project. Declaring a requirement using this method
+	 * instructs the class to initialize it.
+	 *
+	 * Usage example:
+	 * <code>
+	 * $config->setup(
```

Comment:
This feels like a good step in the right direction, but I still think it could be simpler :) (at least for the most common cases)

I would really like things, as much as possible, to have sane defaults and a very very very lightweight initial implementation. Passing in big hashes is a recipe for complex and brittle integrations and it makes it hard to change strings later.

Ideally, in Boost, we would do something like this (pseudocode):

```php
use Automattic\Jetpack\Connection;

static function init() {
  $conn = Connection::init( // invoke the package class directly, rather than via some $config class, which encourages unnecessarily tight coupling between packages
    // $slug - a plugin identifier, defaults to this plugin's slug
    // $name - the plugin name, defaults to this plugin's translated name
    // $file - defaults to this plugin's main file
  );
}
```

Initializing the connection does the following:
- mounts API endpoints for reading and managing the connection
- emits connection state into the page for consumption by UI classes (see below)
- does any necessary plugin registration into a central config, if that's necessary for refcounting the API consumers

And in our UI code:

```jsx
import { ConnectButton, UserMenu, isConnected } from '@automattic/jetpack';

// ...

render() {
  return <>
     { isConnected() && <UserMenu /> }
     // and later
     { ! isConnected() && <ConnectButton /> }
  </>;
}
```

Similarly, if we want to initialize the Sync subsystem, we may invoke:

```php
use Automattic\Jetpack\Sync;
use Automattic\Jetpack\Connection;

static function init() {
  $conn = Connection::init();
  $sync = Sync::init( $conn );
  $sync->register_content_type( 'foo' );
  // or whatever you need to do with Sync
}
```

Now, because Sync has an explicit, initializer-level dependency on Connection (I am assuming here that Sync has a private constructor) it can operate like a "idempotent contextual singleton", giving each invoker a *connection-specific instance* of the Sync class so that operations can be namespaced, checked, etc against the Jetpack connection state of this specific plugin.

Now, let's look at how this would change some existing code, e.g. in the Jetpack plugin:

```php
public function configure() {
		$config = new Config();

		foreach (
			array(
				'sync',
			)
			as $feature
		) {
			$config->ensure( $feature );
		}

		$config->ensure(
			'connection',
			array(
				'slug' => 'jetpack',
				'name' => 'Jetpack',
			)
		);

		if ( is_admin() ) {
			$config->ensure( 'jitm' );
		}

		if ( ! $this->connection_manager ) {
			$this->connection_manager = new Connection_Manager( 'jetpack' );
		}

		/*
		 * Load things that should only be in Network Admin.
		 *
		 * For now blow away everything else until a more full
		 * understanding of what is needed at the network level is
		 * available
		 */
		if ( is_multisite() ) {
			$network = Jetpack_Network::init();
			$network->set_connection( $this->connection_manager );
		}
}
```

This would become:

```php
public function configure() {
		$this->connection= Connection::init();
		$this->sync = Sync::init( $connection );
		$this->connection = JITM::init( $connection );
		// none of the multisite stuff needs to be initialized because it's automatically detected and invoked by the Connection class, but of course you could also initialize it here
}
```

You could also include an additional singleton "get" function which would allow fetching a plugin-specific instance of a class. Effectively, if invoked from Plugin code, it would use the current plugin slug to find and return the correct instance of Sync or whatever for the plugin, and would throw an appropriate exception if the connection hasn't been initialized.

```php
// in some plugin
$jitm = JITM::get(); // throws "Plugin not connected" if it can't find the connection.
$jitm->render_ui(); // or something
```

Under the hood, you could create a centralized instance manager that allows you to manage references to classes, or a base class that contains the appropriate functionality.

For other classes that don't depend on the Connection, but perhaps do depend on each other, you can use a similar set of patterns, I would imagine.

Ultimately, this means we don't actually have (or need) a central entrypoint to configure the Jetpack DNA system as a whole. No "config" class that manages the initialization of all packages centrally. Instead, we have a bunch of coding conventions and attempt to use "zero config" techniques for initializing subsystems as much as possible. This makes this much easier for the plugin embedding the packages.

So, in the case of plugin deactivation, it would be enough just to call `Connection::init()` in most cases and the deactivation hooks will be taken care of for you!

Reply:
Hi @gravityrail.
Thanks for such an extensive review of the approach.

We've been discussing it with the team over the last few days, and are planning provide a proper response later this week.